sycl/test-e2e/Experimental/free_functions/namespace.cpp

```diff
@@ -0,0 +1,235 @@
+#include <sycl/kernel_bundle.hpp>
+#include <sycl/usm.hpp>
+#include <type_traits>
+
+namespace syclext = sycl::ext::oneapi;
+namespace syclexp = sycl::ext::oneapi::experimental;
+
+static constexpr size_t NUM = 1024;
+static constexpr size_t WGSIZE = 16;
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void func(float start, float *ptr) {
+  size_t id = syclext::this_work_item::get_nd_item<1>().get_global_linear_id();
+  ptr[id] = start + static_cast<float>(id);
+}
+
+namespace free_functions::tests {
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void function_in_ns(float start, float *ptr) {
+  size_t id = syclext::this_work_item::get_nd_item<1>().get_global_linear_id();
+  ptr[id] = start + static_cast<float>(id);
+}
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void func(float start, float *ptr) {
+  size_t id = syclext::this_work_item::get_nd_item<1>().get_global_linear_id();
+  ptr[id] = start + static_cast<float>(id);
+}
+} // namespace free_functions::tests
+
+namespace free_functions::tests {
+inline namespace V1 {
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void function_in_inline_ns(float start, float *ptr) {
+  size_t id = syclext::this_work_item::get_nd_item<1>().get_global_linear_id();
+  ptr[id] = start + static_cast<float>(id);
+}
+} // namespace V1
+} // namespace free_functions::tests
+
+namespace {
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void function_in_anonymous_ns(float start, float *ptr) {
+  size_t id = syclext::this_work_item::get_nd_item<1>().get_global_linear_id();
+  ptr[id] = start + static_cast<float>(id);
+}
+} // namespace
+
+struct TestClass {
+  float data;
+  TestClass(float d) : data(d) {}
+};
+
+template <typename T> struct TemplatedTestClass {
+  T data;
+  TemplatedTestClass(T d) : data(d) {}
+};
+
+using IntClassAlias = TemplatedTestClass<int>;
+using FloatClassAlias = TemplatedTestClass<float>;
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void function_with_test_class(float start, TestClass *ptr) {
+  size_t id = syclext::this_work_item::get_nd_item<1>().get_global_linear_id();
+  ptr[id].data = start + static_cast<float>(id);
+}
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void function_with_int_alias_test_class(float start, IntClassAlias *ptr) {
+  size_t id = syclext::this_work_item::get_nd_item<1>().get_global_linear_id();
+  ptr[id].data = start + static_cast<int>(id);
+}
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void function_with_float_alias_test_class(float start, FloatClassAlias *ptr) {
+  size_t id = syclext::this_work_item::get_nd_item<1>().get_global_linear_id();
+  ptr[id].data = start + static_cast<float>(id);
+}
+
+template <typename T>
+concept NumericType = std::is_arithmetic_v<std::remove_reference_t<T>>;
+
+template <typename T>
+  requires NumericType<T>
```

Comment:
I don't know if we allow unconditional C++20 in tests. I remember someone from Codeplay coming to me saying they have runners with old gcc on it. On the other hand, this is core compiler feature, not STL, so maybe nobody would complain.